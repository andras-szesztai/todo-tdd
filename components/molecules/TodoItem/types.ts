export interface TTodoItem {
    id: string
    createdAt: string
    title: string
    isCompleted: boolean
}

export interface Props extends Pick<TTodoItem, 'title' | 'isCompleted'> {
    onCompleteClick: () => void
}
