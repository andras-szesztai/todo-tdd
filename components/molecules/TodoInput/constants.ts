export const TODO_INPUT_LABEL = 'Todo Title'
export const TODO_INPUT_PLACEHOLDER = 'Enter a Todo title...'
export const TODO_INPUT_BUTTON_TEXT = 'Submit'

export const TODO_INPUT_ID = 'todo-input'
