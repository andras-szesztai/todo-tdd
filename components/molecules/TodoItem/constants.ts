export const TODO_COMPLETE_BUTTON_TEXT = {
    complete: 'Complete',
    unComplete: 'Un-complete',
}
