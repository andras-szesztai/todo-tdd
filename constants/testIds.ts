export const TEST_ID = {
    todoInputLabel: 'todo-input-label',
    todoInput: 'todo-input',
    todoInputButton: 'todo-input-button',
    todoViewContainer: 'todos-view-container',
    todoListContainer: 'todo-list-container',
    todoItem: 'todo-item',
    todoCompleteButton: 'todo-complete-button',
}
