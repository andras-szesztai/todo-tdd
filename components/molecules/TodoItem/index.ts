import TodoItem from './TodoItem'
import { TTodoItem } from './types'
import { TODO_COMPLETE_BUTTON_TEXT } from './constants'

export { TodoItem, TODO_COMPLETE_BUTTON_TEXT }
export type { TTodoItem }
