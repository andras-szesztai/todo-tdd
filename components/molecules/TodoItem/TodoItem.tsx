import classNames from 'classnames'
import { Button } from 'components/atoms/Button'
import { TEST_ID } from 'constants/testIds'
import { TODO_COMPLETE_BUTTON_TEXT } from './constants'
import { Props } from './types'

function TodoItem({ title, onCompleteClick, isCompleted }: Props) {
    const cxs = classNames(
        'relative border border-primary-dark rounded-lg p-2  truncate group',
        {
            'bg-transparent border-primary-dark-disabled text-white':
                isCompleted,
            'bg-white border-primary-dark text-primary-dark': !isCompleted,
        }
    )
    return (
        <li data-testid={TEST_ID.todoItem} className={cxs}>
            {title}
            <span className="absolute right-2 top-1/2 -translate-y-1/2 opacity-0 group-hover:opacity-100 transition-opacity">
                <Button
                    testId={TEST_ID.todoCompleteButton}
                    text={
                        isCompleted
                            ? TODO_COMPLETE_BUTTON_TEXT.unComplete
                            : TODO_COMPLETE_BUTTON_TEXT.complete
                    }
                    onClick={onCompleteClick}
                    size="small"
                />
            </span>
        </li>
    )
}

export default TodoItem
