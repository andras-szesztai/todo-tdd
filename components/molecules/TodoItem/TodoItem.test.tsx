import { fireEvent, render, within } from '@testing-library/react'
import { TEST_ID } from 'constants/testIds'

import TodoItem from './TodoItem'

describe('TodoItem', () => {
    it('renders without error', () => {
        const testTitle = 'My Todo'
        const { getByTestId, asFragment } = render(
            <TodoItem
                title={testTitle}
                onCompleteClick={jest.fn}
                isCompleted={false}
            />
        )
        const todoItem = getByTestId(TEST_ID.todoItem)
        expect(todoItem).toBeInTheDocument()
        expect(todoItem).toHaveTextContent(testTitle)
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders Complete button visibly hidden when not hovered', async () => {
        const mockOnCompleteClick = jest.fn()
        const { getByTestId } = render(
            <TodoItem
                title="My Todo"
                onCompleteClick={mockOnCompleteClick}
                isCompleted
            />
        )
        const todoItem = getByTestId(TEST_ID.todoItem)
        const completeTodoButton = within(todoItem).getByTestId(
            TEST_ID.todoCompleteButton
        )
        expect(completeTodoButton.closest('span')).toHaveClass(
            'opacity-0 group-hover:opacity-100'
        )
        expect(mockOnCompleteClick).not.toHaveBeenCalled()
        fireEvent.click(completeTodoButton)
        expect(mockOnCompleteClick).toHaveBeenCalledTimes(1)
    })

    it('renders completed item visually distinct from uncompleted items', () => {
        const uncompletedClasses =
            'bg-white border-primary-dark text-primary-dark'
        const completedClasses =
            'bg-transparent border-primary-dark-disabled text-white'
        const { rerender, getByTestId } = render(
            <TodoItem
                title="My Todo"
                onCompleteClick={jest.fn}
                isCompleted={false}
            />
        )
        const uncompletedTodoItem = getByTestId(TEST_ID.todoItem)
        expect(uncompletedTodoItem).toHaveClass(uncompletedClasses)
        expect(uncompletedTodoItem).not.toHaveClass(completedClasses)
        rerender(
            <TodoItem title="My Todo" onCompleteClick={jest.fn} isCompleted />
        )
        const completedTodoItem = getByTestId(TEST_ID.todoItem)
        expect(uncompletedTodoItem).not.toHaveClass(uncompletedClasses)
        expect(completedTodoItem).toHaveClass(completedClasses)
    })
})
