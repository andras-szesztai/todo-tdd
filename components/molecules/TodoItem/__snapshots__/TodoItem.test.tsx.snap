// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem renders without error 1`] = `
<DocumentFragment>
  <li
    class="relative border border-primary-dark rounded-lg p-2  truncate group bg-white border-primary-dark text-primary-dark"
    data-testid="todo-item"
  >
    My Todo
    <span
      class="absolute right-2 top-1/2 -translate-y-1/2 opacity-0 group-hover:opacity-100 transition-opacity"
    >
      <button
        class="inline text-white font-medium bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent text-xs py-1 px-1.5 rounded-md"
        data-testid="todo-complete-button"
        type="button"
      >
        Complete
      </button>
    </span>
  </li>
</DocumentFragment>
`;
